### ISSUE TITLE
Prim's MST implementation very slow

### ISSUE THREAD

**Issue by Pascal-Ortiz** [networkx/networkx#3502]
Prim's MST implementation in NetworkX is very slow in comparison to Kruskal's one. The performance ratio is pretty high: the higher is the number of nodes, the bigger is the ratio. For instance 

```python
from random import randrange
from time import perf_counter
import networkx as nx

n = 4000
p = 0.25
G = nx.erdos_renyi_graph(n, p)

for e in G.edges():
    G.edges[e]["weight"] = randrange(1, 10)

start = perf_counter()
T = nx.minimum_spanning_tree(G, algorithm='prim')
prim = sum(T.edges[e]["weight"] for e in T.edges())
prim_time=perf_counter() - start
print("Prim : %.2fs" %prim_time)

start = perf_counter()
T = nx.minimum_spanning_tree(G, algorithm='kruskal')
kruskal = sum(T.edges[e]["weight"] for e in T.edges())
kruskal_time=perf_counter() - start
print("Kruskal : %.2fs" %kruskal_time)

print("Ratio: %.1f" %(prim_time/kruskal_time))

# checking
print("Check:", prim == kruskal)
```

outputting
```bash
Prim : 123.26s
Kruskal : 6.55s
Ratio: 18.8
Check: True
```
 

Looking at the Prim's algorithm source code (`mst.py` file), I found 5 spots possibly causing slow execution (cf. the `prim_mst_edges` function).

Here is the first one

```python
    nodes = list(G)
    c = count()

    sign = 1 if minimum else -1

    while nodes:
        u = nodes.pop(0)
```

At the last line, the `pop(0)` remove is not free: `nodes` data structure is a list hence removing at the front has linear cost and since this operation is done inside a `while` loop over the whole graph, this leads to a **quadratic slowdown**.

Later in the code and still inside the `while` loop, we can spot the following:

```python
            nodes.remove(v)
```
            
A `remove` operation on a list has linear complexity so the above line leads again to quadratic slowdown.

And finally, we can spot:

```python
v in visited
```

(this occured 3 times in the code) and since `visited` is a list, this leads again to quadratic slowdown.

To fix the issue, it suffices to change every list to a **set** data structure. I have done some experiments and this causes a nice speedup,  execution is now close to Kruskal's (ratio around 1.2)

**Comment by dschult** [comment]
I don't see any reason not to make the lists into sets. Thanks!

I suspect there is also a speed-up to splitting the large loop into one for MultiGraphs and one for Graphs, rather than a single loop with lots of checking whether G is a multigraph. But I haven't run any tests specifically for this.

**Comment by dschult** [comment]
Closed by #3512
